# Análisis exploratorio de datos de anuncios de autos

## Objetivo

El objetivo de este análisis es explorar y preparar un conjunto de datos de anuncios de vehículos usados para identificar patrones, valores faltantes, valores atípicos y características relacionadas con el precio de los vehículos.

Durante el análisis se realizará la limpieza y preparación de los datos, así como diferentes visualizaciones para comprender la distribución de las principales variables y sus relaciones.

In [204]:
import pandas as pd
import plotly.express as px

## 1. Carga y exploración inicial de los datos

Primero se carga el conjunto de datos y se revisan sus dimensiones, columnas, tipos de datos y valores faltantes.

In [205]:
df = pd.read_csv('../vehicles_us.csv')

df.head()

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
0,9400,2011.0,bmw x5,good,6.0,gas,145000.0,automatic,SUV,NaN,1.0,2018-06-23,19
1,25500,NaN,ford f-150,good,6.0,gas,88705.0,automatic,pickup,white,1.0,2018-10-19,50
2,5500,2013.0,hyundai sonata,like new,4.0,gas,110000.0,automatic,sedan,red,NaN,2019-02-07,79
3,1500,2003.0,ford f-150,fair,8.0,gas,NaN,automatic,pickup,NaN,NaN,2019-03-22,9
4,14900,2017.0,chrysler 200,excellent,4.0,gas,80903.0,automatic,sedan,black,NaN,2019-04-02,28


In [206]:
df.shape

(51525, 13)

In [207]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  str    
 3   condition     51525 non-null  str    
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  str    
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  str    
 8   type          51525 non-null  str    
 9   paint_color   42258 non-null  str    
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  str    
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), str(7)
memory usage: 7.7 MB


## 2. Limpieza y preparación de los datos

Se identificaron valores faltantes en las columnas `model_year`, `cylinders`, `odometer`, `paint_color` e `is_4wd`.

Los valores faltantes se tratarán utilizando métodos adecuados según el significado de cada variable.

In [208]:
df.isna().sum()

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

In [209]:
(df.isna().sum() / len(df)) * 100

price            0.000000
model_year       7.023775
model            0.000000
condition        0.000000
cylinders       10.208637
fuel             0.000000
odometer        15.316836
transmission     0.000000
type             0.000000
paint_color     17.985444
is_4wd          50.369723
date_posted      0.000000
days_listed      0.000000
dtype: float64

In [210]:
df['is_4wd'].value_counts(dropna=False)

is_4wd
NaN    25953
1.0    25572
Name: count, dtype: int64

In [211]:
df['is_4wd'] = df['is_4wd'].fillna(0)

df['is_4wd'].value_counts(dropna=False)

is_4wd
0.0    25953
1.0    25572
Name: count, dtype: int64

In [212]:
df[['model', 'model_year']].head(20)

,model,model_year
0,bmw x5,2011.0
1,ford f-150,NaN
2,hyundai sonata,2013.0
3,ford f-150,2003.0
4,chrysler 200,2017.0
5,chrysler 300,2014.0
6,toyota camry,2015.0
7,honda pilot,2013.0
8,kia sorento,2012.0
9,honda pilot,2008.0


In [213]:
df.groupby('model')['model_year'].median().head(20)

model
acura tl                         2007.0
bmw x5                           2010.0
buick enclave                    2012.0
cadillac escalade                2009.0
chevrolet camaro                 2013.0
chevrolet camaro lt coupe 2d     2017.0
chevrolet colorado               2015.0
chevrolet corvette               2000.0
chevrolet cruze                  2014.0
chevrolet equinox                2013.0
chevrolet impala                 2010.0
chevrolet malibu                 2012.0
chevrolet silverado              2008.0
chevrolet silverado 1500         2011.0
chevrolet silverado 1500 crew    2016.0
chevrolet silverado 2500hd       2010.0
chevrolet silverado 3500hd       2013.0
chevrolet suburban               2008.0
chevrolet tahoe                  2009.0
chevrolet trailblazer            2005.0
Name: model_year, dtype: float64

In [214]:
df['model_year'] = df['model_year'].fillna(
    df.groupby('model')['model_year'].transform('median')
)

df['model_year'].isna().sum()

np.int64(0)

In [215]:
df['model_year'] = df['model_year'].astype(int)

print("Tipo de dato:", df['model_year'].dtype)

df[['model', 'model_year']].head(10)

Tipo de dato: int64


,model,model_year
0,bmw x5,2011
1,ford f-150,2011
2,hyundai sonata,2013
3,ford f-150,2003
4,chrysler 200,2017
5,chrysler 300,2014
6,toyota camry,2015
7,honda pilot,2013
8,kia sorento,2012
9,honda pilot,2008


In [216]:
df['cylinders'].value_counts(dropna=False).sort_index()

cylinders
3.0        34
4.0     13864
5.0       272
6.0     15700
8.0     15844
10.0      549
12.0        2
NaN      5260
Name: count, dtype: int64

In [217]:
df.groupby('model')['cylinders'].median().head(20)

model
acura tl                         6.0
bmw x5                           6.0
buick enclave                    6.0
cadillac escalade                8.0
chevrolet camaro                 6.0
chevrolet camaro lt coupe 2d     6.0
chevrolet colorado               5.0
chevrolet corvette               8.0
chevrolet cruze                  4.0
chevrolet equinox                4.0
chevrolet impala                 6.0
chevrolet malibu                 4.0
chevrolet silverado              8.0
chevrolet silverado 1500         8.0
chevrolet silverado 1500 crew    8.0
chevrolet silverado 2500hd       8.0
chevrolet silverado 3500hd       8.0
chevrolet suburban               8.0
chevrolet tahoe                  8.0
chevrolet trailblazer            6.0
Name: cylinders, dtype: float64

In [218]:
df['cylinders'] = df['cylinders'].fillna(
    df.groupby('model')['cylinders'].transform('median')
)

df['cylinders'].isna().sum()

np.int64(0)

In [219]:
df['cylinders'] = df['cylinders'].astype(int)

df['cylinders'].dtype

dtype('int64')

In [220]:
df['odometer'].describe()

count     43633.000000
mean     115553.461738
std       65094.611341
min           0.000000
25%       70000.000000
50%      113000.000000
75%      155000.000000
max      990000.000000
Name: odometer, dtype: float64

In [221]:
df.groupby('model_year')['odometer'].median().head(20)

model_year
1908    169328.0
1929         NaN
1936     30000.0
1948      4000.0
1949      1800.0
1954      3565.0
1955     47180.0
1958     32991.5
1960     16000.0
1961     66000.0
1962     72000.0
1963     40487.0
1964     58000.0
1965     51330.5
1966     63070.0
1967     94000.0
1968     31362.0
1969     42509.5
1970     83890.0
1971     56329.0
Name: odometer, dtype: float64

In [222]:
df['odometer'] = df['odometer'].fillna(
    df.groupby('model_year')['odometer'].transform('median')
)

df['odometer'].isna().sum()

np.int64(1)

In [223]:
df['odometer'] = df['odometer'].fillna(
    df['odometer'].median()
)

df['odometer'].isna().sum()

np.int64(0)

In [224]:
df['paint_color'].value_counts(dropna=False)

paint_color
white     10029
NaN        9267
black      7692
silver     6244
grey       5037
blue       4475
red        4421
green      1396
brown      1223
custom     1153
yellow      255
orange      231
purple      102
Name: count, dtype: int64

In [225]:
df['paint_color'] = df['paint_color'].fillna('unknown')

df['paint_color'].value_counts(dropna=False)

paint_color
white      10029
unknown     9267
black       7692
silver      6244
grey        5037
blue        4475
red         4421
green       1396
brown       1223
custom      1153
yellow       255
orange       231
purple       102
Name: count, dtype: int64

In [226]:
df.isna().sum()

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

## 3. Corrección de tipos de datos

Algunas columnas tenían tipos de datos que no representaban correctamente su contenido. Se realizaron conversiones para facilitar el análisis posterior.

In [227]:
df['odometer'] = df['odometer'].astype(int)

df['odometer'].dtype

dtype('int64')

In [228]:
df['is_4wd'] = df['is_4wd'].astype(int)

df['is_4wd'].value_counts()

is_4wd
0    25953
1    25572
Name: count, dtype: int64

In [229]:
df['date_posted'] = pd.to_datetime(df['date_posted'])

df['date_posted'].dtype

dtype('<M8[us]')

In [230]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   price         51525 non-null  int64         
 1   model_year    51525 non-null  int64         
 2   model         51525 non-null  str           
 3   condition     51525 non-null  str           
 4   cylinders     51525 non-null  int64         
 5   fuel          51525 non-null  str           
 6   odometer      51525 non-null  int64         
 7   transmission  51525 non-null  str           
 8   type          51525 non-null  str           
 9   paint_color   51525 non-null  str           
 10  is_4wd        51525 non-null  int64         
 11  date_posted   51525 non-null  datetime64[us]
 12  days_listed   51525 non-null  int64         
dtypes: datetime64[us](1), int64(6), str(6)
memory usage: 7.2 MB


## 4. Comprobación de registros duplicados

Se verificó la existencia de filas completamente duplicadas en el conjunto de datos.

In [231]:
df.duplicated().sum()

np.int64(0)

No se encontraron registros completamente duplicados en el conjunto de datos.

### Resumen estadístico de variables numéricas

Antes de analizar las variables individualmente, se revisaron sus principales estadísticas descriptivas para identificar rangos, valores centrales y posibles valores atípicos.

In [232]:
df[['price', 'model_year', 'cylinders', 'odometer', 'days_listed']].describe()

,price,model_year,cylinders,odometer,days_listed
count,51525.000000,51525.000000,51525.000000,51525.000000,51525.00000
mean,12132.464920,2009.793557,6.121067,115199.271771,39.55476
std,10040.803015,6.099381,1.657457,62082.383589,28.20427
min,1.000000,1908.000000,3.000000,0.000000,0.00000
25%,5000.000000,2007.000000,4.000000,73500.000000,19.00000
50%,9000.000000,2011.000000,6.000000,114074.000000,33.00000
75%,16839.000000,2014.000000,8.000000,152827.000000,53.00000
max,375000.000000,2019.000000,12.000000,990000.000000,271.00000


## 5. Análisis inicial de la variable `price`

Se revisó la distribución de los precios para identificar posibles valores atípicos que pudieran afectar las estadísticas y visualizaciones posteriores.

Durante esta primera revisión se detectó una cantidad considerable de anuncios con un precio de $1, valor que no parece representar un precio de venta real para los vehículos observados. Estos registros se analizan y tratan antes de continuar con una revisión más detallada de la variable.

In [233]:
df['price'].sort_values().head(20)

12186    1
40638    1
14644    1
49385    1
11298    1
11297    1
11296    1
11295    1
16868    1
16872    1
29030    1
11320    1
11319    1
11318    1
11317    1
14649    1
41228    1
14647    1
14646    1
14645    1
Name: price, dtype: int64

In [234]:
(df['price'] == 1).sum()

np.int64(798)

In [235]:
df = df[df['price'] != 1]

df.shape

(50727, 13)

In [236]:
print("Registros con precio de $1 después de la limpieza:", (df['price'] == 1).sum())

Registros con precio de $1 después de la limpieza: 0


Se identificaron 798 anuncios con un precio de $1. Debido a la frecuencia de este valor y a que no representa un precio razonable de venta para los vehículos observados, estos registros fueron excluidos del análisis.

In [237]:
df.sort_values('price', ascending=False).head(20)

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
12504,375000,1999,nissan frontier,good,6,gas,115000,automatic,pickup,blue,1,2018-05-19,21
11359,300000,2015,ram 2500,excellent,6,diesel,78285,automatic,truck,grey,1,2018-10-15,39
27375,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2018-09-25,72
34389,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2019-02-02,28
1668,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,unknown,0,2019-03-20,21
33434,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2019-02-05,102
1309,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2019-03-02,56
30634,189000,2014,ford f-150,good,6,gas,90000,automatic,truck,black,0,2018-07-21,42
34206,175000,2004,gmc sierra 2500hd,good,8,diesel,149000,automatic,truck,grey,1,2018-08-25,57
41748,145000,2008,toyota tundra,like new,8,gas,140000,automatic,truck,red,1,2018-06-16,29


### Valores atípicos en los precios

Se analizaron los percentiles de la variable `price` para conocer la distribución de los precios y detectar valores extremadamente altos.

In [238]:
df['price'].quantile([0.90, 0.95, 0.99, 0.995, 0.999])

0.900    25500.000
0.950    30500.000
0.990    43999.000
0.995    49900.000
0.999    64590.818
Name: price, dtype: float64

El 99.9% de los anuncios tiene un precio aproximado de $64,591 o menor. Sin embargo, existen algunos anuncios con precios considerablemente superiores.

Estos registros se conservarán en el conjunto de datos, pero se tendrá en cuenta su efecto al realizar las visualizaciones.

In [239]:
df[df['price'] > 100000].sort_values('price', ascending=False)

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
12504,375000,1999,nissan frontier,good,6,gas,115000,automatic,pickup,blue,1,2018-05-19,21
11359,300000,2015,ram 2500,excellent,6,diesel,78285,automatic,truck,grey,1,2018-10-15,39
1309,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2019-03-02,56
1668,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,unknown,0,2019-03-20,21
27375,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2018-09-25,72
33434,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2019-02-05,102
34389,189000,2014,ford f-150,good,6,gas,151248,automatic,truck,black,0,2019-02-02,28
30634,189000,2014,ford f-150,good,6,gas,90000,automatic,truck,black,0,2018-07-21,42
34206,175000,2004,gmc sierra 2500hd,good,8,diesel,149000,automatic,truck,grey,1,2018-08-25,57
41748,145000,2008,toyota tundra,like new,8,gas,140000,automatic,truck,red,1,2018-06-16,29


## 6. Análisis exploratorio mediante visualizaciones

### Distribución de precios

Primero se observa la distribución completa de los precios.

In [240]:
fig = px.histogram(
    df,
    x='price',
    title='Distribución de precios de vehículos'
)

fig.show()

La distribución presenta una fuerte asimetría hacia la derecha. Algunos precios extremadamente altos amplían considerablemente el eje horizontal y dificultan observar la distribución de la mayoría de los vehículos.

In [241]:
limite_price_99 = df['price'].quantile(0.99)

fig = px.histogram(
    df[df['price'] <= limite_price_99],
    x='price',
    nbins=50,
    title='Distribución de precios de vehículos - 99% de los datos'
)

fig.show()

print("Límite del percentil 99:", limite_price_99)

Límite del percentil 99: 43999.0


Al observar la distribución completa se aprecia una fuerte asimetría hacia la derecha debido a algunos vehículos con precios extremadamente altos.

Para visualizar con mayor claridad la concentración principal de los datos, se realizó una segunda gráfica limitada al percentil 99 del precio. Este filtro se utilizó únicamente para la visualización y no modifica el conjunto de datos original.

La mayor concentración de anuncios se encuentra en los rangos de precios bajos y medios, mientras que la frecuencia disminuye conforme aumenta el precio.

In [242]:
df['model_year'].describe()

count    50727.000000
mean      2009.706527
std          6.080772
min       1908.000000
25%       2007.000000
50%       2011.000000
75%       2014.000000
max       2019.000000
Name: model_year, dtype: float64

### Análisis del año del modelo

Se analizó la distribución de `model_year` para conocer los años de los vehículos publicados e identificar posibles valores atípicos.

La mediana del año del modelo es 2011 y el 50% de los vehículos se encuentra aproximadamente entre 2007 y 2014. También se identificaron algunos registros considerablemente más antiguos, por lo que se realizó una revisión específica de estos casos antes de decidir su tratamiento.

In [243]:
df['model_year'].value_counts().sort_index().head(20)

model_year
1908     2
1929     1
1936     1
1948     1
1949     1
1954     1
1955     1
1958     2
1960     3
1961     1
1962     3
1963     3
1964     7
1965    10
1966    22
1967     7
1968    16
1969    10
1970    10
1971     7
Name: count, dtype: int64

In [244]:
df[df['model_year'] < 1960].sort_values('model_year')

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
33906,12995,1908,gmc yukon,good,8,gas,169328,automatic,SUV,black,0,2018-07-06,34
33907,12995,1908,cadillac escalade,excellent,8,gas,169328,automatic,SUV,white,0,2018-06-24,25
45694,18000,1929,ford f-150,good,8,gas,114074,manual,other,silver,0,2018-11-18,59
34713,5000,1936,ford f-150,excellent,6,gas,30000,manual,pickup,purple,0,2018-11-22,10
22595,21000,1948,chevrolet impala,like new,8,gas,4000,automatic,sedan,red,0,2019-01-18,24
36582,44900,1949,chevrolet suburban,good,8,gas,1800,automatic,wagon,orange,0,2018-08-19,10
14752,15000,1954,ford f-150,excellent,8,gas,3565,manual,pickup,black,0,2019-02-16,13
10018,23900,1955,ford f250,excellent,6,gas,47180,manual,truck,blue,0,2018-12-22,61
39580,35000,1958,chevrolet impala,excellent,8,gas,3184,automatic,coupe,black,0,2018-05-19,33
48414,37900,1958,chevrolet impala,good,8,gas,62799,automatic,coupe,unknown,0,2018-08-11,10


In [245]:
(df['model_year'] < 1960).sum()

np.int64(10)

In [246]:
vehiculos_antiguos = df[df['model_year'] < 1960]

print("Vehículos anteriores a 1960:", len(vehiculos_antiguos))
print("Porcentaje del dataset:", round(len(vehiculos_antiguos) / len(df) * 100, 2), "%")

Vehículos anteriores a 1960: 10
Porcentaje del dataset: 0.02 %


In [337]:
df.loc[df['model_year'] < 1960, 'model_year'].value_counts().sort_index()

model_year
1908    2
1929    1
1936    1
1948    1
1949    1
1954    1
1955    1
1958    2
Name: count, dtype: int64

### Tratamiento de vehículos antiguos

Se identificaron 10 registros con años de modelo anteriores a 1960, equivalentes aproximadamente al 0.02% del conjunto de datos.

Debido a que un año de modelo antiguo no constituye por sí solo evidencia suficiente de un error, estos registros no se eliminaron automáticamente. Algunos pueden corresponder a vehículos clásicos o de colección.

Sin embargo, la revisión también mostró casos en los que el año del modelo podría no ser coherente con el vehículo anunciado. Por esta razón, estos registros se conservaron por el momento y se tendrán en cuenta al interpretar los análisis posteriores relacionados con el año del modelo, la antigüedad y el precio.

In [345]:
print("Millaje igual a 0:", (df['odometer'] == 0).sum())
print("Millaje menor a 100:", (df['odometer'] < 100).sum())
print("Millaje mayor a 300,000:", (df['odometer'] > 300000).sum())
print("Millaje mayor a 500,000:", (df['odometer'] > 500000).sum())

Millaje igual a 0: 0
Millaje menor a 100: 165
Millaje mayor a 300,000: 287
Millaje mayor a 500,000: 22


In [250]:
df['odometer'].quantile([0.90, 0.95, 0.99, 0.995, 0.999])

0.900    190000.000
0.950    217000.000
0.990    280375.680
0.995    306805.960
0.999    395112.614
Name: odometer, dtype: float64

In [251]:
df[df['odometer'] < 100].sort_values(
    ['odometer', 'model_year']
).head(30)

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
21083,20000,1962,chevrolet impala,good,6,gas,0,automatic,sedan,blue,0,2019-02-14,57
7517,10500,1966,chevrolet impala,fair,8,gas,0,automatic,convertible,yellow,0,2019-03-16,27
33893,6250,1968,chrysler 300,good,8,gas,0,automatic,coupe,green,0,2018-10-01,9
2049,30000,1969,chevrolet corvette,excellent,8,other,0,automatic,convertible,white,0,2018-08-01,20
27609,49000,1969,chevrolet camaro,like new,8,gas,0,manual,coupe,custom,0,2019-03-12,40
23025,4500,1970,ford f-250,good,8,gas,0,automatic,pickup,white,0,2018-09-04,21
28799,45900,1971,chevrolet camaro,like new,6,gas,0,manual,coupe,orange,0,2019-01-10,75
30251,2750,1984,chevrolet silverado 2500hd,fair,8,gas,0,automatic,truck,blue,0,2018-11-20,54
30191,2500,1988,ford f-150,good,8,gas,0,automatic,pickup,unknown,1,2018-10-05,22
37252,1290,1988,ford f-150,good,8,gas,0,manual,pickup,brown,1,2019-02-23,53


In [252]:
df[df['odometer'] > 500000].sort_values(
    'odometer',
    ascending=False
)

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
40729,4700,2013,chevrolet cruze,good,6,gas,990000,automatic,sedan,black,0,2018-05-02,4
17869,59900,1964,chevrolet corvette,like new,8,gas,990000,automatic,convertible,red,0,2018-06-17,28
48962,40000,2014,ram 3500,excellent,6,diesel,980000,automatic,truck,white,1,2018-12-06,7
14245,6000,1996,ram 2500,good,10,gas,920000,automatic,truck,blue,0,2019-02-02,43
2840,18500,2014,ford explorer,excellent,6,gas,866000,automatic,SUV,black,1,2018-07-20,14
2440,5300,2010,toyota corolla,excellent,4,gas,840000,automatic,sedan,silver,0,2019-02-04,13
32973,5550,2010,toyota corolla,excellent,4,gas,840000,automatic,sedan,silver,0,2018-10-21,20
13027,8500,2015,chevrolet malibu,excellent,4,gas,838095,automatic,sedan,silver,0,2018-05-30,10
41699,10300,2013,ford escape,excellent,6,gas,831000,automatic,SUV,brown,0,2018-05-27,30
5735,7500,2015,hyundai elantra,excellent,4,gas,813542,other,coupe,brown,0,2018-11-21,5


In [253]:
df.loc[df['odometer'] == 0, 'odometer'] = pd.NA

df['odometer'] = df['odometer'].fillna(
    df.groupby('model_year')['odometer'].transform('median')
)

df['odometer'] = df['odometer'].fillna(
    df['odometer'].median()
)

df['odometer'] = df['odometer'].astype(int)

In [346]:
print("Valores nulos:", df['odometer'].isna().sum())
print("Millaje igual a 0:", (df['odometer'] == 0).sum())
print("Tipo de dato:", df['odometer'].dtype)

Valores nulos: 0
Millaje igual a 0: 0
Tipo de dato: int64


### Tratamiento de valores atípicos en `odometer`

Se analizaron los valores extremos de millaje. Se identificaron registros con millaje igual a 0, los cuales se consideraron valores faltantes o incorrectos, especialmente al observarse en vehículos con varios años de antigüedad.

Estos valores se sustituyeron utilizando la mediana del millaje correspondiente al mismo año del modelo (`model_year`), ya que la antigüedad del vehículo está relacionada con el millaje esperado. Como medida adicional, cualquier valor faltante restante se completó con la mediana general.

También se identificaron vehículos con millaje extremadamente altos. Sin embargo, estos registros representan una proporción muy pequeña del conjunto de datos y no existe evidencia suficiente para considerarlos errores, por lo que se conservaron.

In [347]:
print("Anuncios con 0 días:", (df['days_listed'] == 0).sum())
print("Más de 100 días:", (df['days_listed'] > 100).sum())
print("Más de 150 días:", (df['days_listed'] > 150).sum())
print("Más de 200 días:", (df['days_listed'] > 200).sum())

Anuncios con 0 días: 54
Más de 100 días: 1917
Más de 150 días: 236
Más de 200 días: 33


In [348]:
df['days_listed'].quantile([0.90, 0.95, 0.99, 0.995, 0.999])

0.900     77.0
0.950     94.0
0.990    132.0
0.995    149.0
0.999    187.0
Name: days_listed, dtype: float64

### Análisis de `days_listed`

Se analizó el número de días que permanecieron publicados los anuncios para identificar posibles valores atípicos.

La mayoría de los anuncios permanece publicada durante periodos relativamente cortos: el 90% dura 77 días o menos, el 95% hasta 94 días y el 99% hasta 132 días.

Se encontraron 54 anuncios con `days_listed` igual a 0. Estos valores se conservaron, ya que pueden representar anuncios retirados el mismo día de su publicación o registros con una duración inferior a un día. También existen algunos anuncios con periodos superiores a 200 días, pero representan una proporción muy pequeña del conjunto de datos y son valores posibles.

Por lo tanto, no se realizaron modificaciones en esta variable.

In [ ]:
print("Dimensiones:", df.shape) ------------ 717171717171
print("\nValores nulos:")
print(df.isna().sum())

print("\nDuplicados:", df.duplicated().sum())

print("\nTipos de datos:")
print(df.dtypes)

Dimensiones: (50727, 13)

Valores nulos:
price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

Duplicados: 0

Tipos de datos:
price                    int64
model_year               int64
model                      str
condition                  str
cylinders                int64
fuel                       str
odometer                 int64
transmission               str
type                       str
paint_color                str
is_4wd                   int64
date_posted     datetime64[us]
days_listed              int64
dtype: object


## Conclusión de la limpieza de datos

Después del proceso de preprocesamiento, el conjunto de datos contiene 50,727 registros y 13 variables.

Se trataron los valores ausentes utilizando criterios adecuados para cada variable, se corrigieron tipos de datos y se revisaron posibles valores atípicos e inconsistencias. También se eliminaron 798 registros con precio igual a $1, al considerarse valores no representativos para el análisis de precios.

La comprobación final confirmó que no existen valores nulos ni registros duplicados y que las variables cuentan con tipos de datos adecuados para continuar con el análisis.

A partir de este punto se utilizará el conjunto de datos limpio para estudiar los factores relacionados con el precio y el comportamiento de los anuncios de vehículos.

# Análisis exploratorio de datos (EDA)

Una vez finalizada la limpieza de los datos, se analizarán las principales características de los vehículos y su relación con el precio.

El objetivo de esta etapa es identificar patrones, tendencias y relaciones entre variables como el año del modelo, kilometraje, condición, tipo de vehículo, transmisión y otras características que puedan influir en el precio de venta.

In [258]:
print("Precio promedio:", round(df['price'].mean(), 2))
print("Precio mediano:", df['price'].median())
print("Precio mínimo:", df['price'].min())
print("Precio máximo:", df['price'].max())

Precio promedio: 12323.31
Precio mediano: 9495.0
Precio mínimo: 3
Precio máximo: 375000


In [259]:
df['price'].value_counts().sort_index().head(30)

price
3       1
5       1
6       1
9       1
10      1
11      1
12      3
15      3
20      1
24      1
25      2
28      1
32      1
35      2
36      1
39      3
65      1
69     34
80      2
85      4
105     1
111     1
147     1
155     3
169     1
171     1
176     3
179     1
180     2
185     3
Name: count, dtype: int64

In [260]:
print("Menores de $100:", (df['price'] < 100).sum())
print("Menores de $500:", (df['price'] < 500).sum())
print("Menores de $1,000:", (df['price'] < 1000).sum())

Menores de $100: 65
Menores de $500: 271
Menores de $1,000: 606


In [261]:
df[df['price'] < 100].sort_values('price')

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
44205,3,2005,jeep liberty,good,6,gas,153108,automatic,SUV,unknown,1,2018-06-19,22
50430,5,2011,toyota sienna,good,6,gas,123025,automatic,SUV,green,0,2018-12-03,5
31263,6,1999,ford f250,good,8,gas,173500,automatic,pickup,unknown,1,2019-02-07,53
39379,9,2010,subaru forester,good,4,gas,119,automatic,SUV,grey,1,2019-03-02,12
50971,10,2012,toyota prius,excellent,4,hybrid,101000,automatic,hatchback,green,0,2019-03-16,29
...,...,...,...,...,...,...,...,...,...,...,...,...,...
29129,80,2005,toyota camry le,good,4,gas,145033,automatic,sedan,silver,0,2019-01-11,25
15151,85,2012,ford fusion se,good,4,gas,110000,automatic,sedan,white,0,2019-01-30,57
11840,85,2007,toyota corolla,good,4,gas,110850,automatic,sedan,silver,0,2018-11-20,48
11784,85,2012,ram 1500,good,8,gas,133691,automatic,truck,unknown,0,2018-10-16,160


In [262]:
print("Menos de $100:", (df['price'] < 100).sum())
print("$100 - $249:", ((df['price'] >= 100) & (df['price'] < 250)).sum())
print("$250 - $499:", ((df['price'] >= 250) & (df['price'] < 500)).sum())
print("$500 - $749:", ((df['price'] >= 500) & (df['price'] < 750)).sum())
print("$750 - $999:", ((df['price'] >= 750) & (df['price'] < 1000)).sum())

Menos de $100: 65
$100 - $249: 90
$250 - $499: 116
$500 - $749: 156
$750 - $999: 179


In [263]:
df = df[df['price'] >= 100].copy()

print("Dimensiones:", df.shape)
print("Precio mínimo:", df['price'].min())
print("Registros con precio menor a $100:", (df['price'] < 100).sum())

Dimensiones: (50662, 13)
Precio mínimo: 105
Registros con precio menor a $100: 0


### Tratamiento de valores atípicos en `price`

Durante el análisis de la variable `price` se identificaron valores extremadamente bajos que no resultaban coherentes con las características de los vehículos anunciados. Por ejemplo, se encontraron vehículos en condiciones buenas o excelentes con precios inferiores a $100.

Se analizaron distintos rangos de precios bajos antes de realizar modificaciones. Debido a que los precios entre $100 y $1,000 pueden corresponder a vehículos antiguos, dañados o de bajo valor, se decidió conservarlos para evitar eliminar observaciones potencialmente válidas.

En cambio, se eliminaron los 65 registros con precios inferiores a $100. Estos representan una proporción mínima del conjunto de datos y, tras revisar sus características, se consideraron valores suficientemente atípicos como para afectar el análisis de la distribución de precios.

In [264]:
df['price'].describe()

count     50662.000000
mean      12339.051834
std        9999.350084
min         105.000000
25%        5300.000000
50%        9495.000000
75%       16950.000000
max      375000.000000
Name: price, dtype: float64

In [265]:
print("Precio promedio:", round(df['price'].mean(), 2))
print("Precio mediano:", df['price'].median())
print("Diferencia promedio-mediana:", round(df['price'].mean() - df['price'].median(), 2))

Precio promedio: 12339.05
Precio mediano: 9495.0
Diferencia promedio-mediana: 2844.05


### Distribución general de los precios

Después de la limpieza, el precio promedio de los vehículos es de aproximadamente $12,339, mientras que la mediana es de $9,495.

La diferencia entre ambas medidas muestra que la distribución de precios presenta un sesgo hacia la derecha. Esto se debe a la presencia de vehículos con precios considerablemente altos, que elevan el promedio.

El 50% central de los precios se encuentra aproximadamente entre $5,300 y $16,950. Debido al sesgo de la distribución, la mediana resulta una medida más representativa del precio típico de los vehículos del conjunto de datos.

In [266]:
fig = px.histogram(
    df,
    x='price',
    nbins=100,
    title='Distribución de precios de vehículos'
)

fig.show()

In [267]:
limite_99 = df['price'].quantile(0.99)

fig = px.histogram(
    df[df['price'] <= limite_99],
    x='price',
    nbins=50,
    title='Distribución de precios de vehículos - 99% de los datos'
)

fig.show()

print("Límite del percentil 99:", limite_99)

Límite del percentil 99: 43999.0


### Distribución de `price`

La distribución de los precios presenta un claro sesgo hacia la derecha. La mayor parte de los vehículos se concentra en los rangos de precios más bajos, mientras que la frecuencia disminuye progresivamente conforme aumenta el precio.

La existencia de un pequeño número de vehículos con precios muy elevados genera una cola larga hacia la derecha y provoca que el precio promedio ($12,339) sea superior a la mediana ($9,495).

Para visualizar con mayor claridad el comportamiento de la mayoría de los anuncios, se realizó una segunda gráfica limitada al percentil 99, correspondiente a un precio de $43,999. Este filtro se aplicó únicamente a la visualización y no modifica el conjunto de datos.

In [350]:
precio_por_año = (
    df.groupby('model_year')['price']
    .agg(['count', 'median'])
    .reset_index()
)

precio_por_año.tail(15)

,model_year,count,median
53,2005,2065,4995.0
54,2006,2271,5900.0
55,2007,2831,6495.0
56,2008,3451,7488.0
57,2009,2411,6900.0
58,2010,3185,7990.0
59,2011,4001,10563.0
60,2012,3953,9900.0
61,2013,3969,11900.0
62,2014,3468,13995.0


In [351]:
fig = px.line(
    precio_por_año,
    x='model_year',
    y='median',
    markers=True,
    title='Precio mediano por año del modelo'
)

fig.show()

In [352]:
fig = px.line(
    precio_por_año[precio_por_año['model_year'] >= 1990],
    x='model_year',
    y='median',
    markers=True,
    title='Precio mediano por año del modelo (1990-2019)'
)

fig.show()

### Relación entre el año del modelo y el precio

Al analizar el precio mediano según el año del modelo se observa una tendencia general en la que los vehículos más recientes presentan precios medianos más altos.

Los años más antiguos muestran mayores fluctuaciones y cuentan con una menor cantidad de observaciones, por lo que sus medianas deben interpretarse con cautela.

Al enfocar la visualización en los modelos de 1990 a 2019 se aprecia con mayor claridad esta tendencia, especialmente a partir de los años 2000. En general, los resultados sugieren una relación positiva entre el año del modelo y el precio: los vehículos de años más recientes tienden a presentar precios más altos.

In [353]:
fig = px.scatter(
    df,
    x='odometer',
    y='price',
    opacity=0.3,
    title='Relación entre millaje y precio'
)

fig.show()

In [354]:
limite_odometer_99 = df['odometer'].quantile(0.99)
limite_price_99 = df['price'].quantile(0.99)

df_scatter = df[
    (df['odometer'] <= limite_odometer_99) &
    (df['price'] <= limite_price_99)
]

fig = px.scatter(
    df_scatter,
    x='odometer',
    y='price',
    opacity=0.3,
    title='Relación entre millaje y precio - 99% de los datos'
)

fig.show()

print("Límite odometer percentil 99:", limite_odometer_99)
print("Límite price percentil 99:", limite_price_99)

Límite odometer percentil 99: 280468.5
Límite price percentil 99: 43999.0


### Relación entre millaje y precio

La visualización general muestra que existen algunos valores extremos tanto en `odometer` como en `price`, los cuales dificultan observar con claridad la concentración principal de los datos.

Por esta razón, se realizó una segunda visualización limitada al percentil 99 de ambas variables. El 99% de los registros presenta un millaje de hasta aproximadamente 280,469 millas y un precio de hasta $43,999. Este filtro se utilizó únicamente para mejorar la visualización y no modifica el conjunto de datos original.

Al observar la concentración principal de los datos se aprecia una relación negativa entre el millaje y el precio. En general, conforme aumenta el millaje, los vehículos tienden a presentar precios más bajos. Sin embargo, existe una dispersión considerable, especialmente entre los vehículos con menor millaje, lo que sugiere que otras características también están relacionadas con el precio.

In [273]:
precio_por_condicion = (
    df.groupby('condition')['price']
    .agg(['count', 'median'])
    .sort_values('median')
)

precio_por_condicion

,count,median
condition,,
fair,1597,2500.0
salvage,115,2500.0
good,20112,7900.0
excellent,23987,10900.0
like new,4712,13995.0
new,139,22300.0


In [274]:
fig = px.bar(
    precio_por_condicion.reset_index(),
    x='condition',
    y='median',
    text='median',
    title='Precio mediano según la condición del vehículo'
)

fig.show()

### Relación entre condición y precio

El precio mediano presenta una tendencia creciente conforme mejora la condición del vehículo. Los vehículos clasificados como `fair` y `salvage` presentan una mediana de $2,500, mientras que las categorías `good`, `excellent` y `like new` alcanzan medianas de $7,900, $10,900 y $13,995, respectivamente.

Los vehículos clasificados como `new` presentan el precio mediano más alto, con $22,300. Sin embargo, esta categoría contiene únicamente 139 registros, por lo que debe interpretarse con mayor cautela que categorías con un número considerablemente mayor de observaciones.

En general, los resultados muestran una asociación clara entre la condición del vehículo y el precio mediano: las categorías que representan mejores condiciones tienden a presentar precios más altos.

In [275]:
precio_por_tipo = (
    df.groupby('type')['price']
    .agg(['count', 'median'])
    .sort_values('median')
)

precio_por_tipo

,count,median
type,,
hatchback,1036,5995.0
sedan,11983,5995.0
mini-van,1159,6500.0
wagon,1541,7900.0
van,616,7995.0
SUV,12135,8995.0
other,255,8995.0
bus,24,10500.0
offroad,212,11800.0


In [276]:
fig = px.bar(
    precio_por_tipo.reset_index(),
    x='type',
    y='median',
    text='median',
    title='Precio mediano según el tipo de vehículo'
)

fig.show()

### Relación entre tipo de vehículo y precio

El precio mediano presenta diferencias importantes según el tipo de vehículo. Los `hatchback` y `sedan` muestran los precios medianos más bajos, ambos con $5,995, mientras que los `pickup` y `truck` presentan los valores más altos, con $14,400 y $15,300, respectivamente.

También se observan precios medianos relativamente altos en categorías como `coupe`, `convertible` y `offroad`. Estas diferencias muestran que el precio mediano varía de forma considerable entre los distintos tipos de vehículo.

Sin embargo, algunas categorías cuentan con pocas observaciones, como `bus`, por lo que sus resultados deben interpretarse con mayor cautela.

In [277]:
precio_por_transmision = (
    df.groupby('transmission')['price']
    .agg(['count', 'median'])
    .sort_values('median')
)

precio_por_transmision

,count,median
transmission,,
manual,2788,6999.0
automatic,46362,9100.0
other,1512,21900.0


In [278]:
fig = px.bar(
    precio_por_transmision.reset_index(),
    x='transmission',
    y='median',
    text='median',
    title='Precio mediano según el tipo de transmisión'
)

fig.show()

### Relación entre transmisión y precio

Se observan diferencias en el precio mediano según el tipo de transmisión. Los vehículos con transmisión `manual` presentan el precio mediano más bajo, con $6,999, mientras que los vehículos `automatic` alcanzan una mediana de $9,100.

La categoría `other` presenta un precio mediano considerablemente mayor, de $21,900. Sin embargo, esta categoría cuenta con menos observaciones y puede estar asociada con otras características de los vehículos.

En general, los resultados muestran una asociación entre el tipo de transmisión y el precio mediano, aunque estas diferencias deben interpretarse considerando otras variables como el tipo de vehículo, el año del modelo y el millaje.

In [279]:
precio_por_combustible = (
    df.groupby('fuel')['price']
    .agg(['count', 'median'])
    .sort_values('median')
)

precio_por_combustible

,count,median
fuel,,
hybrid,407,6650.0
electric,5,7900.0
gas,46437,8980.0
other,108,16493.0
diesel,3705,19995.0


In [280]:
fig = px.bar(
    precio_por_combustible.reset_index(),
    x='fuel',
    y='median',
    text='median',
    title='Precio mediano según el tipo de combustible'
)

fig.show()

### Relación entre combustible y precio

Se observan diferencias en el precio mediano según el tipo de combustible. Los vehículos `hybrid` presentan una mediana de $6,650, los vehículos de `gas` de $8,980 y los vehículos `diesel` alcanzan la mediana más alta, con $19,995.

La categoría `other` también presenta un precio mediano relativamente alto, de $16,493. Por otro lado, la categoría `electric` cuenta únicamente con 5 registros, por lo que su mediana de $7,900 debe interpretarse con especial cautela.

En general, el precio mediano varía entre los distintos tipos de combustible. Sin embargo, estas diferencias pueden estar asociadas con otras características de los vehículos, como el tipo, el año del modelo, la antigüedad o el millaje.

### Antigüedad del vehículo

Además del año del modelo, se creó la variable `vehicle_age` para representar la antigüedad que tenía cada vehículo en el momento en que se publicó el anuncio.

Esta variable se calculó como la diferencia entre el año de publicación del anuncio (`date_posted`) y el año del modelo (`model_year`). De esta forma, es posible analizar de manera más directa la relación entre la antigüedad del vehículo y su precio.

In [281]:
df['vehicle_age'] = df['date_posted'].dt.year - df['model_year']

In [282]:
df[['date_posted', 'model_year', 'vehicle_age']].head(10)

,date_posted,model_year,vehicle_age
0,2018-06-23,2011,7
1,2018-10-19,2011,7
2,2019-02-07,2013,6
3,2019-03-22,2003,16
4,2019-04-02,2017,2
5,2018-06-20,2014,4
6,2018-12-27,2015,3
7,2019-01-07,2013,6
8,2018-07-16,2012,6
9,2019-02-15,2008,11


In [283]:
df['vehicle_age'].describe()

count    50662.000000
mean         8.604181
std          6.093482
min          0.000000
25%          4.000000
50%          8.000000
75%         12.000000
max        110.000000
Name: vehicle_age, dtype: float64

In [284]:
print("Edades negativas:", (df['vehicle_age'] < 0).sum())
print("Edad igual a 0:", (df['vehicle_age'] == 0).sum())
print("Edad mayor a 50:", (df['vehicle_age'] > 50).sum())

Edades negativas: 0
Edad igual a 0: 1878
Edad mayor a 50: 71


In [356]:
precio_por_antiguedad = (
    df.groupby('vehicle_age')['price']
    .agg(['count', 'median'])
    .reset_index()
)

precio_por_antiguedad

,vehicle_age,count,median
0,0,1878,25629.0
1,1,1984,21000.0
2,2,2756,19700.0
3,3,3142,18490.0
4,4,3460,14950.0
...,...,...,...
63,69,1,44900.0
64,71,1,21000.0
65,82,1,5000.0
66,89,1,18000.0


In [357]:
fig = px.line(
    precio_por_antiguedad,
    x='vehicle_age',
    y='median',
    markers=True,
    title='Precio mediano según antigüedad del vehículo'
)

fig.show()

In [358]:
precio_por_antiguedad.head(10)

,vehicle_age,count,median
0,0,1878,25629.0
1,1,1984,21000.0
2,2,2756,19700.0
3,3,3142,18490.0
4,4,3460,14950.0
5,5,3806,12495.0
6,6,3919,9995.0
7,7,4045,10000.0
8,8,3388,8750.0
9,9,2716,6995.0


In [359]:
precio_por_antiguedad.tail(10)

,vehicle_age,count,median
58,58,2,42500.0
59,59,1,17500.0
60,60,2,36450.0
61,63,1,23900.0
62,65,1,15000.0
63,69,1,44900.0
64,71,1,21000.0
65,82,1,5000.0
66,89,1,18000.0
67,110,2,12995.0


### Relación entre antigüedad y precio

Se observa una relación negativa entre la antigüedad del vehículo y su precio, especialmente entre los vehículos de menor antigüedad. Los vehículos nuevos o con pocos años presentan precios medianos considerablemente mayores, mientras que el precio disminuye progresivamente conforme aumenta `vehicle_age`.

Por ejemplo, los vehículos con 0 años de antigüedad presentan un precio mediano de $25,629, mientras que los de 5 años tienen una mediana de $12,495 y los de 10 años de $7,495.

En los vehículos con mayor antigüedad se observan fluctuaciones importantes y, en algunos casos, un aumento del precio mediano. Sin embargo, estas edades cuentan con un número reducido de observaciones y pueden incluir vehículos clásicos o de colección, por lo que estos resultados deben interpretarse con cautela.

En general, la antigüedad es uno de los factores que muestra una relación más clara con el precio: los vehículos más recientes tienden a presentar precios más altos.

In [289]:
limite_edad_99 = df['vehicle_age'].quantile(0.99)

print("Percentil 99 de antigüedad:", limite_edad_99)
print("Vehículos por encima:", (df['vehicle_age'] > limite_edad_99).sum())

Percentil 99 de antigüedad: 28.0
Vehículos por encima: 446


In [360]:
precio_por_antiguedad_99 = precio_por_antiguedad[
    precio_por_antiguedad['vehicle_age'] <= limite_edad_99
]

fig = px.line(
    precio_por_antiguedad_99,
    x='vehicle_age',
    y='median',
    markers=True,
    title='Precio mediano según antigüedad - 99% de los datos'
)

fig.show()

In [361]:
precio_por_antiguedad_99.tail()

,vehicle_age,count,median
24,24,150,3650.0
25,25,99,3495.0
26,26,80,3500.0
27,27,46,3850.0
28,28,62,5500.0


Para observar con mayor claridad esta relación, se realizó una segunda visualización limitada al percentil 99 de `vehicle_age`, correspondiente a vehículos con hasta 27 años de antigüedad. Este filtro se utilizó únicamente para la visualización y no modificó el conjunto de datos.

Al concentrarse en el 99% de las observaciones, la tendencia resulta más evidente: el precio mediano disminuye con rapidez durante los primeros años de antigüedad y posteriormente la reducción se vuelve más gradual. Esto refuerza la relación observada entre la antigüedad del vehículo y su precio.

In [292]:
precio_por_4wd = (
    df.groupby('is_4wd')['price']
    .agg(['count', 'median'])
)

precio_por_4wd

,count,median
is_4wd,,
0,25778,6995.0
1,24884,13900.0


In [293]:
fig = px.bar(
    precio_por_4wd.reset_index(),
    x='is_4wd',
    y='median',
    text='median',
    title='Precio mediano según tracción 4WD'
)

fig.show()

## Relación entre tracción 4WD y precio

Se observa una diferencia importante en el precio mediano según la presencia de tracción 4WD. Los vehículos sin tracción 4WD presentan un precio mediano de $6,995, mientras que los vehículos con tracción 4WD alcanzan una mediana de $13,900.

Ambos grupos cuentan con una cantidad similar de observaciones, con 25,778 vehículos sin 4WD y 24,884 vehículos con 4WD, lo que permite realizar una comparación descriptiva entre ambas categorías.

En general, los vehículos con tracción 4WD tienden a presentar precios más altos. Sin embargo, esta diferencia no implica que la tracción 4WD sea por sí sola la causa del mayor precio, ya que también puede estar relacionada con otras características como el tipo de vehículo, el año del modelo, la antigüedad o el millaje.

In [362]:
precio_por_cilindros = (
    df.groupby('cylinders')['price']
    .agg(['count', 'median'])
    .sort_index()
)

precio_por_cilindros

,count,median
cylinders,,
3,33,7850.0
4,15360,6950.0
5,293,6500.0
6,17212,9500.0
8,17509,13990.0
10,253,16900.0
12,2,19000.0


In [363]:
fig = px.bar(
    precio_por_cilindros.reset_index(),
    x='cylinders',
    y='median',
    text='median',
    title='Precio mediano según el número de cilindros'
)

fig.show()

## Relación entre número de cilindros y precio

Se observa una tendencia general de aumento en el precio mediano conforme aumenta el número de cilindros. Entre las categorías con mayor cantidad de observaciones, los vehículos de 4 cilindros presentan una mediana de $6,950, los de 6 cilindros de $9,500 y los de 8 cilindros de $13,990.

Los vehículos de 10 y 12 cilindros presentan medianas aún mayores, de $16,900 y $19,000, respectivamente. Sin embargo, estas categorías cuentan con pocas observaciones, especialmente la de 12 cilindros, que contiene únicamente 2 registros, por lo que sus resultados deben interpretarse con cautela.

En general, el número de cilindros muestra una relación con el precio del vehículo, aunque esta tendencia no se cumple de manera estricta en todas las categorías y puede estar relacionada con otras características como el tipo de vehículo, el combustible o la tracción.

In [296]:
df['days_listed'].describe()

count    50662.000000
mean        39.563440
std         28.208461
min          0.000000
25%         19.000000
50%         33.000000
75%         53.000000
max        271.000000
Name: days_listed, dtype: float64

In [297]:
print("Mediana:", df['days_listed'].median())
print("Anuncios con 0 días:", (df['days_listed'] == 0).sum())
print("Más de 100 días:", (df['days_listed'] > 100).sum())
print("Máximo:", df['days_listed'].max())

Mediana: 33.0
Anuncios con 0 días: 54
Más de 100 días: 1917
Máximo: 271


In [298]:
fig = px.scatter(
    df,
    x='days_listed',
    y='price',
    opacity=0.3,
    title='Relación entre tiempo publicado y precio'
)

fig.show()

In [299]:
limite_days_99 = df['days_listed'].quantile(0.99)
limite_price_99 = df['price'].quantile(0.99)

df_days_price = df[
    (df['days_listed'] <= limite_days_99) &
    (df['price'] <= limite_price_99)
]

fig = px.scatter(
    df_days_price,
    x='days_listed',
    y='price',
    opacity=0.3,
    title='Relación entre tiempo publicado y precio - 99% de los datos'
)

fig.show()

print("Límite days_listed percentil 99:", limite_days_99)
print("Límite price percentil 99:", limite_price_99)

Límite days_listed percentil 99: 132.0
Límite price percentil 99: 43999.0


## Relación entre tiempo publicado y precio

La visualización general presenta algunos precios extremadamente altos que dificultan observar la concentración principal de los datos. Por esta razón, se realizó una segunda visualización limitada al percentil 99 de `days_listed` y `price`. El 99% de los registros presenta un tiempo de publicación de hasta 132 días y un precio de hasta $43,999. Este filtro se utilizó únicamente para mejorar la visualización y no modifica el conjunto de datos original.

Al observar la concentración principal de los datos, no se aprecia una relación fuerte y claramente definida entre el tiempo que permanece publicado un anuncio y el precio del vehículo. Para distintos valores de `days_listed` existe una amplia variedad de precios.

Aunque los precios más altos parecen ser menos frecuentes entre los anuncios que permanecen publicados durante periodos más largos, la dispersión de los datos es considerable. Por lo tanto, no existe evidencia visual suficiente para afirmar que el precio por sí solo determine cuánto tiempo permanece publicado un vehículo.

In [300]:
precio_por_color = (
    df.groupby('paint_color')['price']
    .agg(['count', 'median'])
    .sort_values('median')
)

precio_por_color

,count,median
paint_color,,
purple,102,4997.0
green,1383,5995.0
brown,1215,7300.0
blue,4420,7500.0
silver,6166,8000.0
custom,1082,8497.5
grey,4975,8750.0
red,4370,8900.0
unknown,9096,9500.0


In [301]:
fig = px.bar(
    precio_por_color.reset_index(),
    x='paint_color',
    y='median',
    text='median',
    title='Precio mediano según el color del vehículo'
)

fig.show()

## Relación entre color y precio

Se observan diferencias en el precio mediano según el color del vehículo. Entre las categorías con mayor cantidad de observaciones, los vehículos de colores como `green`, `blue` y `silver` presentan precios medianos relativamente bajos, mientras que `black` y `white` alcanzan medianas mayores, de $11,495 y $11,888, respectivamente.

Los colores `orange` y `yellow` presentan los precios medianos más altos, con $15,700 y $17,490. Sin embargo, estas categorías cuentan con pocas observaciones en comparación con los colores más frecuentes, por lo que sus resultados deben interpretarse con cautela. De forma similar, `purple`, que presenta la mediana más baja de $4,997, cuenta únicamente con 102 registros.

La categoría `unknown` representa los vehículos para los que no se dispone de información sobre el color y presenta una mediana de $9,500.

En general, se observan diferencias de precio entre las categorías de color, aunque no existe evidencia suficiente para considerar que el color por sí solo explique estas diferencias, ya que puede estar relacionado con otras características como el tipo de vehículo, el modelo o la antigüedad.

In [302]:
print("Número de modelos diferentes:", df['model'].nunique())

df['model'].value_counts().head(20)

Número de modelos diferentes: 100


model
ford f-150                    2736
chevrolet silverado 1500      2157
ram 1500                      1715
chevrolet silverado           1259
ram 2500                      1078
jeep wrangler                 1064
toyota camry                   985
honda accord                   954
chevrolet silverado 2500hd     914
gmc sierra 1500                889
ford explorer                  843
ford escape                    839
toyota tacoma                  825
nissan altima                  799
honda civic                    792
jeep grand cherokee            772
ford focus                     753
honda cr-v                     677
chevrolet tahoe                676
chevrolet impala               674
Name: count, dtype: int64

In [ ]:
df.to_csv('vehicles_clean.csv', index=False) ------ 135 --- REVISAR 

In [304]:
precio_por_modelo = (
    df.groupby('model')['price']
    .agg(['count', 'median'])
    .sort_values('count', ascending=False)
)

precio_por_modelo.head(20)

,count,median
model,,
ford f-150,2736,12500.0
chevrolet silverado 1500,2157,13995.0
ram 1500,1715,14850.0
chevrolet silverado,1259,12000.0
ram 2500,1078,20990.0
jeep wrangler,1064,13995.0
toyota camry,985,5900.0
honda accord,954,5495.0
chevrolet silverado 2500hd,914,16995.0


In [305]:
top_20_modelos = precio_por_modelo.head(20).reset_index()

fig = px.bar(
    top_20_modelos,
    x='model',
    y='median',
    text='median',
    title='Precio mediano de los 20 modelos más frecuentes'
)

fig.show()

## Relación entre modelo y precio

Se observan diferencias importantes en el precio mediano entre los 20 modelos con mayor cantidad de observaciones. `ram 2500` presenta la mediana más alta, con $20,990, mientras que `chevrolet impala` presenta la más baja, con $4,700.

En general, modelos de pickups como `ram 2500`, `ram 1500`, `chevrolet silverado 1500` y `toyota tacoma` presentan precios medianos relativamente altos, mientras que automóviles como `honda civic`, `ford focus` y `chevrolet impala` muestran precios menores.

Por lo tanto, el modelo muestra una relación importante con el precio. Sin embargo, estas diferencias también pueden estar asociadas con otras características como el tipo de vehículo, la tracción, el número de cilindros o la antigüedad.

In [ ]:
df['model'] = df['model'].str.upper() ----- 139 ----- REVISAR 

In [ ]:
df.to_csv('vehicles_clean.csv', index=False) ---- 141 ----- REVSAR 

In [334]:
columnas_numericas = [
    'price',
    'model_year',
    'odometer',
    'cylinders',
    'is_4wd',
    'days_listed',
    'vehicle_age'
]

correlaciones = df[columnas_numericas].corr()

correlaciones['price'].sort_values(ascending=False)

price          1.000000
model_year     0.441416
is_4wd         0.334538
cylinders      0.328534
days_listed   -0.001077
vehicle_age   -0.440064
odometer      -0.465232
Name: price, dtype: float64

### Correlación de variables numéricas con el precio

El análisis de correlación confirma varias de las relaciones observadas anteriormente. El millaje (`odometer`) presenta la relación negativa más fuerte con el precio (-0.465), indicando que los vehículos con mayor millaje tienden a presentar precios menores.

El año del vehículo (`model_year`) presenta una correlación positiva de 0.450, mientras que la antigüedad (`vehicle_age`) presenta una correlación negativa de -0.448. Esto indica que los vehículos más recientes tienden a presentar precios mayores.

Las variables `is_4wd` y `cylinders` presentan correlaciones positivas de 0.335 y 0.329, respectivamente. Por otro lado, `days_listed` presenta una correlación cercana a cero (-0.001), por lo que no se observa una relación lineal relevante con el precio.

En general, el millaje, el año y la antigüedad del vehículo muestran las relaciones lineales más importantes con el precio entre las variables numéricas analizadas.

In [336]:
fig = px.histogram(
    df,
    x='days_listed',
    nbins=50,
    title='Distribución de días publicados'
)

fig.show()

### Distribución de días publicados

La mayoría de los anuncios permanece publicada durante aproximadamente 10 a 60 días. La mediana es de 33 días y el promedio de 39.6 días.

La distribución presenta una cola hacia la derecha debido a una pequeña cantidad de anuncios que permanecen publicados durante periodos considerablemente mayores, alcanzando un máximo de 271 días.

En general, un anuncio típico permanece publicado alrededor de un mes, mientras que los periodos superiores a 100 días son poco frecuentes.

## Conclusiones generales del análisis exploratorio

El análisis muestra que el precio de los vehículos está relacionado con varias de sus características. Entre las variables numéricas, el millaje, el año y la antigüedad presentan las relaciones más claras con el precio. Los vehículos más recientes y con menor millaje tienden a presentar precios mayores.

También se observaron diferencias importantes según las características del vehículo. Los vehículos con tracción 4WD y mayor número de cilindros presentan precios medianos más altos. De igual manera, el tipo, la condición, la transmisión y el combustible muestran diferencias en sus precios medianos.

El modelo del vehículo también resulta relevante, ya que se encontraron diferencias importantes entre los modelos más frecuentes del conjunto de datos.

Por otro lado, `days_listed` presenta una correlación prácticamente nula con el precio, por lo que el tiempo que permanece publicado un anuncio no parece tener una relación lineal importante con esta variable.

En conjunto, los resultados indican que el precio no depende de una sola característica, sino de la combinación de factores como el modelo, el año, el millaje, la condición, el tipo de vehículo, la tracción y el número de cilindros.